docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+  - An image can only be upgraded with an image of higher version number.
+  - For each image, which has v1 and v2 version and for each dependent d,
+    if v1 depends on d then v2 also depends on d
+  - For each image, which has v1 and v2 version, if v2 > v1 then for each d
+    where v1 depends on d, v2.dependentversion[d] >= v1.dependentversion[d]
+  - Removing and introducing dependencies with new image version is not allowed,
+    which means the same amount of dependency entry must be in image manifest.
+
+Observations:
+  - There is always at least 1 valid configuration, the initial state must be
+    consistent. In worst case, the system returns to this state after dependency
+    check.
+  - Decreasing the version number of the actual slot either doesn't change,
+    or decrease the number of unsatisfied dependencies.
+  - In case `REVERT` the image, which is taken into account during dependency
+    check has lower version than image in primary slot.
+
+``` c
+
+void *get_dependency_entry(uint32_t image_id, uint32_t slot);
+void  get_next_dependency_entry(void**);
+uint32_t get_dependent_image_id(void *);
+uint64_t get_dependent_image_min_version(void *);
+uint64_t get_image_version(uint32_t image_id, uint32_t slot);
+
+i = 0;
+while (i < IMAGE_CNT)
+{
+    curr_slot = (boot_data[i].swap_type != NONE) ? 1 : 0;
+
+    ptr = get_dependency_entry(i, curr_slot);
+    while(ptr)
+    {
+        id = get_dependent_image_id(ptr);
+        min_version = get_dependent_image_min_version(ptr);
+        get_next_dependency_entry(&ptr);
+
+        dep_slot = (boot_data[id].swap_type != NONE) ? 1 : 0;
+        dep_version = get_image_version(id, dep_slot);
+
+        if (min_version <= dep_version)
+        {
+            /* Dependency satisfied, no action needed */
+            i++;
+        }
+        else
+        {
+            /* Dependency not satisfied */
+            switch (boot_data[i].swap_type)
+            {
+                case PERM:
+                case TEST:
+                    boot_data[i].swap_type = NONE;
+                break;
+                case NONE:
+                    boot_data[i].swap_type = REVERT;
+                break;
+            }
+
+            /* Dependency check needs to be restarted */
+            i = 0;
+            break;
+        }
+    }
+}
+```
+**
+
+Valid states after dependency check:
+  - In a consistent state images should move in the same direction or stay in
+    place. This means in practice that `swap_type` cannot be at the same time
+    `REVERT` and `<PERM | TEST>`.
+  - These `swap_type` combinations are valid:
+    - `NONE` or `<PERM | TEST>`
+    - `REVERT` or `NONE`
+
+# Testing in simulator
+Simulator environment must be improved to able to test possible scenarios during
+multiple image update, focusing on abort and revert scenarios and on the usage
+of common scratch area.
+
+# Testing on hardware
+Testing can be done on ARMv8-M based platform i.e.: MPS2_AN521 or Musca A1.
```

Comment:
Unfortunately, Musca A1 isn't a great target to test this, since it runs out of SRAM. It does have the ability to directly execute out of the serial flash, but this shouldn't be used for anything other than copying mcuboot to RAM. The execution directly from flash runs at about 1/75 regular speed (RSA signature check takes around 10 seconds).

Support for Musca A1 is kind of a separate proposal, and we indeed have other targets that boot from flash loading the code into SRAM.

Qemu is also a reasonable choice. I'm not sure how much has been merged back into mainline, but there is v8m support on qemu.

Reply:
If I got it right then this slow performance happens if MCBUoot is executed from flash. Do you know how much time does it take if MCUBoot copied to SRAM and executed from there, but images still in flash?

I'm checking the status of armV8m support in QEMU.

Reply:
According to Peter Maydell v8m support has merged to master. 

We did measurement on Musca A1, if MCUBoot was executed from SRAM:
 - default setting for QSPI access: 1.9s
 - enhanced QSPI access: 1.6s

The measurement covers the complete MCUBoot execution: hash and signature checking, copying image to SRAM, hash checking again.